### PR TITLE
Risk based speedup algorithm added

### PR DESCRIPTION
**Pull request by ldg810**
In global planner, the setpoint to follow global path is updated periodically at "GlobalPlannerNode::publishSetpoint()"
The speed of drone is keep constantly (2.0 m/s at current situation).
However, it would be efficient to speed up in low-risk environment.

I added following two parameters to apply risk-based speedup algorithm.
- max_speed_(double) : maximum speed of drone in safe environment.
- use_risk_based_speedup_(bool) : boolean to use risk based speedup algorithm

The algorithm is simple
1. get current cell from global_planner
2. get risk of current cell (=cur_risk)
3. if risk is high(>=0.5), set speed to slow (1.0m/s)
    else, set speed to fast (1.0 + (global_planner_.max_speed_ - 1.0) * (1 - cur_risk))
4. isCloseToGoal is also changed to determine the drone is close to goal if distance from destination and drone is within current speed setpoint. (If not, the drone would oscillate in fast flight)

Following graph is comparison between default algorithm(red line) and risk based speedup algorithm(green line) from a point to another point
abcd

### CODE REVIEW

File: global_planner/src/nodes/global_planner_node.cpp [hunk 1]
```diff
@@ -436,6 +438,18 @@ void GlobalPlannerNode::printPointInfo(double x, double y, double z) {
 void GlobalPlannerNode::publishSetpoint() {
   // Vector pointing from current position to the current goal
   tf::Vector3 vec = toTfVector3(subtractPoints(current_goal_.pose.position, last_pos_.pose.position));
+  if (global_planner_.use_speedup_heuristics_){
+    Cell cur_cell = global_planner::Cell(last_pos_.pose.position.x, last_pos_.pose.position.y, last_pos_.pose.position.z);
+    double cur_risk = std::sqrt(global_planner_.getRisk(cur_cell));
+    if(cur_risk >= 0.5) {   // If current risk is too high, set speed as low to stable flight.
+      speed_ = 1.0;
```
Comment (Jaeyoung-Lim) [inline global_planner/src/nodes/global_planner_node.cpp:445]: These numbers look like magic numbers, could you parameterize these numbers somehow so it doesn't get lost in the code?

**Comment by ldg810** [comment]
> Thanks for the contribution!
> 
> A few questions/comments
> a. Could you also resolve the conflict that are shown probably by rebasing?
> b. Could you also maybe post a video on how this is an improvement?
> c. Have you flown this on a real vehicle? or is this solely tested on SITL?

Hi, Thank you for your comments.

a. I resolved conflict by rebasing the code.
b. I upload a movie clip to compare between default global planner and risk based speed up algorithm (https://youtu.be/k2_xxLZEQb0)
c. I didn't fly this algorithm on a real vehicle, it was tested solely on SITL.

File: global_planner/src/nodes/global_planner_node.cpp (hunk 1 above)
Comment (jkflying) [reply global_planner/src/nodes/global_planner_node.cpp:427]: Agreed, in different situations you might have the `global_planner_.max_speed_` to be less than 1. Rather make this a fraction of the max_speed_, and rework the algorithm below to linearly interpolate from there.

Comment (ldg810) [reply global_planner/src/nodes/global_planner_node.cpp]: @Jaeyoung-Lim @jkflying I made the numbers as parameteres (risk_threshold_risk_based_speedup_, default_speed_) and modify speed_ calculation equation accordingly.

**Comment by Jaeyoung-Lim** [comment]
@ldg810 Any chance if this can be flight tested before getting merged?

**Comment by ldg810** [comment]
> @ldg810 Any chance if this can be flight tested before getting merged?
> 
> Also, please look at the CI failures (Travis) the style checks are failing

I think I will be able to test this algorithm with PX4 Vision Kit within this month.
Then, I will try to test it with real flight and share the result in here.

Also, I tried to use fix_style.sh in tools (with clang-format-3.9 since there is no 3.8 version for ubuntu 18.04. Refer : https://pkgs.org/download/clang-format-3.8) but it results in very different format for global_planner/cfg/GlobalPlannerNode.cfg file. (For example, after fixing, "#!/usr/bin/env python" in line 1 changed to "#!/ usr / bin / env / python").
Should I use clang-format-3.8? or any changes required for fix_style.sh?

**Review by jkflying — APPROVED** [review]

**Comment by ldg810** [comment]
> @ldg810 Any chance if this can be flight tested before getting merged?
> 
> Also, please look at the CI failures (Travis) the style checks are failing

Hi, There was two typhoon and heavy rain on my city nowadays.

I tried to test global_planner in soccer field but I have some difficulties to test it in real world.. (It is hard to find good obstacle which is located in big empty field :( )

- WiFi connection between PX4-Vision Kit and laptop is very unreliable so it is hard to monitoring full path plan in real time. (It often disconnected in distance of more than 10 to 15 m).
- Launching global planner in PX4-Vision Kit with automated script at the time of booting cause failure of some subsystem but I was not able to exactly figure out what is the problem. Thus, I should log in to PX4-Vision Kit with ssh and launch global planner, then it works well.
- Today, I acquired rosbag data of flight with global planner so I will analyse the data and will try to improve the planner.

**Comment by Jaeyoung-Lim** [comment]
> WiFi connection between PX4-Vision Kit and laptop is very unreliable so it is hard to monitoring full path plan in real time. (It often disconnected in distance of more than 10 to 15 m).

The wifi module (ESP8266) is not supposed to be something that you use in operation, but more for close proximity operations, such as calibration or tuning the vehicle.

> Launching global planner in PX4-Vision Kit with automated script at the time of booting cause failure of some subsystem but I was not able to exactly figure out what is the problem. Thus, I should log in to PX4-Vision Kit with ssh and launch global planner, then it works well.

Did you write your own automated script?

> Today, I acquired rosbag data of flight with global planner so I will analyse the data and will try to improve the planner.

Awesome! A flight log will be super useful, in addition to the rosbags

**Comment by ldg810** [comment]
> Did you write your own automated script?

Yes, I mixed launch file of mavros / depth camera / global_planner like this.

```
<launch>
    <arg name="ns" default="/"/>
    <arg name="fcu_url" default="/dev/ttyUSB0:921600"/>
    <arg name="gcs_url" default="" />   <!-- GCS link is provided by SITL -->
    <arg name="tgt_system" default="1" />
    <arg name="tgt_component" default="1" />
    <arg name="start_pos_x" default="0.0" />
    <arg name="start_pos_y" default="0.0" />
    <arg name="start_pos_z" default="3.0" />
    
    <!-- Launch MavROS -->
    <group ns="$(arg ns)">
        <include file="$(find mavros)/launch/node.launch">
            <arg name="pluginlists_yaml" value="$(find mavros)/launch/px4_pluginlists.yaml" />
            <!-- Need to change the config file to get the tf topic and get local position in terms of local origin -->
            <arg name="config_yaml" value="$(find global_planner)/resource/px4_config.yaml" />
            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="$(arg gcs_url)" />
            <arg name="tgt_system" value="$(arg tgt_system)" />
            <arg name="tgt_component" value="$(arg tgt_component)" />
        </include>
    </group>

    <node pkg="nodelet" type="nodelet" name="auterion_nodelet_manager" args="manager" output="screen" required="true"/>
    <node pkg="tf" type="static_transform_publisher" name="tf_camera_main" required="true"
       args="0.1 0.0 -0.03 -1.571 0.0 -1.571 fcu camera_main_map 10"/>

    <!-- Launch Depth Camera -->
    <include file="$(find structure_core_ros_driver)/launch/sc.launch">
       <arg name="name"                  value="camera_main"/>
       <arg name="required"              value="true"/>
       <arg name="serial_number"         value="null"/>
       <arg name="infrared_enable"       value="false" />
       <arg name="depth_enable"          value="true" />
       <arg name="depth_pcloud_enable"   value="true" />
       <arg name="frame_sync_enabled"    value="false" />
       <arg name="depth_framerate"       value="15" />
       <arg name="depth_resolution"      value="_320x240" />
       <arg name="depth_apply_correction_before_stream"              value="true"/>
    </include>

    <!-- Launch Global Planner -->
    <arg name="pointcloud_topics" default="[/camera_main/depth/points]"/>
    <include file="$(find global_planner)/launch/global_planner_octomap.launch" >
        <arg name="pointcloud_topics" value="$(arg pointcloud_topics)" />
        <arg name="start_pos_x" value="$(arg start_pos_x)" />
        <arg name="start_pos_y" value="$(arg start_pos_y)" />
        <arg name="start_pos_z" value="$(arg start_pos_z)" />
    </include>
</launch>
```

and i modified the avoidance.service script to launch global_planner instead of local_planner
In current situation, the service(enabled) return "missing required data" with automated start with booting, but I restart(sudo systemctl restart avoidance.service) the service, then it works well.

**Pull request merged**